apollo-router-core/src/services/checkpoint/async_checkpoint.rs

```diff
@@ -0,0 +1,294 @@
+use super::Step;
+use futures::future::BoxFuture;
+use std::sync::Arc;
+use tower::{
+    buffer::Buffer, util::BoxCloneService, BoxError, Layer, Service, ServiceBuilder, ServiceExt,
+};
+
+#[allow(clippy::type_complexity)]
+pub struct AsyncCheckpointLayer<S, Request>
+where
+    S: Service<Request, Error = BoxError> + Clone + Send + 'static,
+    Request: Send + 'static,
+    S::Future: Send,
+    S::Response: Send + 'static,
+{
+    checkpoint_fn: Arc<
+        dyn Fn(
+                Request,
+            ) -> BoxFuture<
+                'static,
+                Result<Step<Request, <S as Service<Request>>::Response>, BoxError>,
+            > + Send
+            + Sync
+            + 'static,
+    >,
+}
+
+#[allow(clippy::type_complexity)]
+impl<S, Request> AsyncCheckpointLayer<S, Request>
+where
+    S: Service<Request, Error = BoxError> + Clone + Send + 'static,
+    Request: Send + 'static,
+    S::Future: Send,
+    S::Response: Send + 'static,
+{
+    /// Create an `AsyncCheckpointLayer` from a function that takes a Service Request and returns a `Step`
+    pub fn new(
+        checkpoint_fn: impl Fn(
+                Request,
+            ) -> BoxFuture<
+                'static,
+                Result<Step<Request, <S as Service<Request>>::Response>, BoxError>,
+            > + Send
+            + Sync
+            + 'static,
+    ) -> Self {
+        Self {
+            checkpoint_fn: Arc::new(checkpoint_fn),
+        }
+    }
+}
+
+impl<S, Request> Layer<S> for AsyncCheckpointLayer<S, Request>
+where
+    S: Service<Request, Error = BoxError> + Clone + Send + 'static,
+    <S as Service<Request>>::Future: Send,
+    Request: Send + 'static,
+    <S as Service<Request>>::Response: Send + 'static,
+{
+    type Service = AsyncCheckpointService<
+        BoxCloneService<Request, <S as Service<Request>>::Response, BoxError>,
+        Request,
+    >;
+
+    fn layer(&self, service: S) -> Self::Service {
+        let inner = Buffer::new(service, 20_000);
+        AsyncCheckpointService {
+            checkpoint_fn: Arc::clone(&self.checkpoint_fn),
+            inner: ServiceBuilder::new().service(inner).boxed_clone(),
```

Comment:
Should just be able to store inner directly here.

Reply:
Also need to remove the buffer.

Reply:
nice catch it s a leftover indeed!